File: Swift2/Sources/Swift2/TrigramValueSequencer.swift

public class CTrigramValueSequencer : CValueSequencer {
	
	init(nValue: Int) {
		super.init(nInnerSequencers: 3, nValues: 8, nValue: nValue)
		m_pvsInner = [CLineValueSequencer(nValue: 0) as CValueSequencer,CLineValueSequencer(nValue: 0) as CValueSequencer,
			CLineValueSequencer(nValue: 0) as CValueSequencer]
		//m_pvsInner[2] = new CLineValueSequencer(0)
		//m_pvsInner[1] = new CLineValueSequencer(0)
		//m_pvsInner[0] = new CLineValueSequencer(0)
		Line(nIndex: 2).SetParent(pvsParent: self)
		Line(nIndex: 1).SetParent(pvsParent: self)
		Line(nIndex: 0).SetParent(pvsParent: self)
		m_nSequences = Sequences.nTrigramSequences
		m_nRatios = Sequences.nTrigramRatios
	}

	public func Line(nIndex: Int) -> CLineValueSequencer {
 		return m_pvsInner[nIndex] as! CLineValueSequencer
	}

	@discardableResult
	public override func Inverse() -> CValueSequencer {
		var bMoving: [Bool] = [ false, false, false ]
		SaveMoving(bMoving: &bMoving)
		Value = (Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) +
			((Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			((Line(nIndex: 0).Value % 2 == 0 ? 0 : 1) * 4)
		UpdateInnerValues()
		UpdateOuterValues()
		RestoreMoving(bMoving: &bMoving, bInverseLine: true)
		return self
	}

	@discardableResult
	public override func Opposite() -> CValueSequencer {
		var bMoving: [Bool] = [ false, false, false ]
		SaveMoving(bMoving: &bMoving)
		super.Opposite()
		RestoreMoving(bMoving: &bMoving, bInverseLine: false)
		return self
	}

	@discardableResult
	public override func Move() -> CValueSequencer {
		Line(nIndex: 2).Move()
		Line(nIndex: 1).Move()
		Line(nIndex: 0).Move()
		return self
	}

	@discardableResult
	public override func Young() -> CValueSequencer {
		Line(nIndex: 2).Young()
		Line(nIndex: 1).Young()
		Line(nIndex: 0).Young()
		return self
	}

	public override func UpdateInnerValues() {
		switch Value {
		case 0:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYinLine
			break
		case 1:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYangLine
			break
		case 2:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYinLine
			break
		case 3:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYangLine
			break
		case 4:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYinLine
			break
		case 5:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYinLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYangLine
			break
		case 6:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYinLine
			break
		case 7:
			Line(nIndex: 2).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 1).Value = CTrigramValueSequencer.m_nYangLine
			Line(nIndex: 0).Value = CTrigramValueSequencer.m_nYangLine
			break
		default:
			let _ = 0
		}
		Line(nIndex: 2).UpdateInnerValues()
		Line(nIndex: 1).UpdateInnerValues()
		Line(nIndex: 0).UpdateInnerValues()
	}

	public override func UpdateOuterValues() {
		Value = (Line(nIndex: 0).Value % 2 == 0 ? 0 : 1) +
			((Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			((Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) * 4)
		m_pvsParent?.UpdateOuterValues()
	}

	private func SaveMoving(bMoving: inout [Bool]) {
		for l in 0 ... 2 {
			if (Line(nIndex: l).Value == 0) || (Line(nIndex: l).Value == 3) {
				bMoving[l] = true
			}
		}
	}

	private func RestoreMoving(bMoving: inout [Bool], bInverseLine: Bool) {
		for l in 0 ... 2 {
			if bMoving[l]	{
				var l1 = l
				if bInverseLine {
					if l1 == 0 {
						l1 = 2
					}
					else
					if l1 == 2 {
						l1 = 0
					}
				Line(nIndex: l1).Old()
				}
			}
		}
	}

	public static func SetCurrentSequence(nSequence: Int) {
		m_nCurrentSequence = nSequence
	}

	public static func SetCurrentRatio(nRatio: Int) {
		m_nCurrentRatio = nRatio
	}

	public static func SetCurrentLabel(nLabel: Int) {
		m_nCurrentLabel = nLabel
	}

	public override func GetLabel() -> String { 
		return Sequences.strTrigramLabels[GetCurrentLabel()][Value]
	}

	public override func GetMoving() -> Bool {
		return Line(nIndex: 0).IsMoving || Line(nIndex: 1).IsMoving || Line(nIndex: 2).IsMoving
	}

	public override func GetCurrentSequence() -> Int { return CTrigramValueSequencer.m_nCurrentSequence }
	public override func GetCurrentRatio() -> Int { return CTrigramValueSequencer.m_nCurrentRatio }
	public override func GetCurrentLabel() -> Int{ return CTrigramValueSequencer.m_nCurrentLabel }

	private static var m_nCurrentSequence: Int = 0
	private static var m_nCurrentRatio: Int = 0
	private static var m_nCurrentLabel: Int = 0

	private static var m_nYinLine: Int = 2
	private static var m_nYangLine: Int = 1
}

File: ValueSequencer.py/trigram_value_sequencer.py
from __future__ import annotations

from typing import List

from .line_value_sequencer import CLineValueSequencer
from .value_sequencer import CValueSequencer
from .sequences import Sequences


class CTrigramValueSequencer(CValueSequencer):
    _nCurrentSequence = 0
    _nCurrentRatio = 0
    _nCurrentLabel = 0

    def __init__(self, value: int) -> None:
        super().__init__(3, 8, value)
        self._pvsInner = [CLineValueSequencer(0) for _ in range(3)]
        for line in self._pvsInner:
            line.set_parent(self)
        self._nSequences = Sequences.nTrigramSequences
        self._nRatios = Sequences.nTrigramRatios
        self._nYinLine = 2
        self._nYangLine = 1

    def line(self, index: int) -> CLineValueSequencer:
        return self._pvsInner[index]  # type: ignore[return-value]

    def inverse(self) -> "CTrigramValueSequencer":
        moving = self._save_moving()
        self.value = (
            (self.line(2).value % 2) +
            (self.line(1).value % 2) * 2 +
            (self.line(0).value % 2) * 4
        )
        self.update_inner_values()
        self.update_outer_values()
        self._restore_moving(moving, True)
        return self

    def opposite(self) -> "CTrigramValueSequencer":
        moving = self._save_moving()
        super().opposite()
        self._restore_moving(moving, False)
        return self

    def move(self) -> "CTrigramValueSequencer":
        for idx in range(3):
            self.line(idx).move()
        return self

    def young(self) -> "CTrigramValueSequencer":
        for idx in range(3):
            self.line(idx).young()
        return self

    def update_inner_values(self) -> None:
        configs: List[List[int]] = [
            [self._nYinLine, self._nYinLine, self._nYinLine],
            [self._nYinLine, self._nYinLine, self._nYangLine],
            [self._nYinLine, self._nYangLine, self._nYinLine],
            [self._nYinLine, self._nYangLine, self._nYangLine],
            [self._nYangLine, self._nYinLine, self._nYinLine],
            [self._nYangLine, self._nYinLine, self._nYangLine],
            [self._nYangLine, self._nYangLine, self._nYinLine],
            [self._nYangLine, self._nYangLine, self._nYangLine],
        ]
        values = configs[self.value]
        for index, line in enumerate(reversed(self._pvsInner)):
            line.value = values[index]
            line.update_inner_values()

    def update_outer_values(self) -> None:
        self.value = (
            (self.line(0).value % 2) +
            (self.line(1).value % 2) * 2 +
            (self.line(2).value % 2) * 4
        )
        if self._pvsParent:
            self._pvsParent.update_outer_values()

    @classmethod
    def set_current_sequence(cls, sequence: int) -> None:
        cls._nCurrentSequence = sequence

    @classmethod
    def set_current_ratio(cls, ratio: int) -> None:
        cls._nCurrentRatio = ratio

    @classmethod
    def set_current_label(cls, label: int) -> None:
        cls._nCurrentLabel = label

    def get_label(self) -> str:
        return Sequences.strTrigramLabels[self.get_current_label()][self.value]

    def get_moving(self) -> bool:
        return any(self.line(idx).is_moving for idx in range(3))

    def get_current_sequence(self) -> int:
        return CTrigramValueSequencer._nCurrentSequence

    def get_current_ratio(self) -> int:
        return CTrigramValueSequencer._nCurrentRatio

    def get_current_label(self) -> int:
        return CTrigramValueSequencer._nCurrentLabel

    def _save_moving(self) -> List[bool]:
        moving = [False, False, False]
        for idx in range(3):
            line = self.line(idx)
            if line.value in {0, 3}:
                moving[idx] = True
        return moving

    def _restore_moving(self, moving: List[bool], inverse_line: bool) -> None:
        for idx, is_moving in enumerate(moving):
            if not is_moving:
                continue
            target = idx
            if inverse_line:
                if target == 0:
                    target = 2
                elif target == 2:
                    target = 0
            self.line(target).old()
